homeassistant/components/buienradar/camera.py

```diff
@@ -0,0 +1,176 @@
+    _name = ""              # type: str
+    _delta = 0.0         # type: float
+    _condition = None       # type: Optional[asyncio.Condition]
+    """ Loading status """
+    _loading = False        # type: bool
+
+    _last_image = None      # type: Optional[bytes]
+    """ last modified HTTP response header"""
+    _last_modified = None   # type: Optional[str]
+
+    def __init__(self, hass, name: str, c_id: Optional[str], dimension: int,
+                 delta: float):
+        """Initialize the component."""
+        super().__init__()
+        self._hass = hass
+        self._name = name
+
+        self._dimension = dimension
+        self._delta = delta
+
+        self._condition = asyncio.Condition(loop=hass.loop)
+
+        self._last_image = None
+        self._last_modified = None
+        self._loading = False
+        # deadline for image to be refreshed
+        self._deadline = None
+
+    @property
+    def name(self) -> str:
+        """Return the component name."""
+        return self._name
+
+    # Cargo-cult from components/proxy/camera.py
+    def camera_image(self) -> Optional[bytes]:
+        """Return camera image."""
+        return run_coroutine_threadsafe(
+            self.async_camera_image(), self.hass.loop).result()
+
+    def __needs_refresh(self) -> bool:
+        if not (self._delta and self._deadline and self._last_image):
+            return True
+
+        return dt_util.utcnow() > self._deadline
+
+    async def __retrieve_radar_image(self) -> bool:
+        """Retrieve new radar image and return whether this succeeded."""
+        session = async_get_clientsession(self._hass)
+
+        url = RADAR_MAP_URL_TEMPLATE.format(w=self._dimension,
+                                            h=self._dimension)
+
+        if self._last_modified:
+            headers = {'If-Modified-Since': self._last_modified}
+        else:
+            headers = {}
+
+        try:
+            async with session.get(url, timeout=5, headers=headers) as res:
+                if res.status == 304:
+                    _LOG.debug("HTTP 304 - success")
+                    return True
+                if res.status != 200:
+                    _LOG.error("HTTP %s - failure", res.status)
+                    return False
+
+                last_modified = res.headers.get('Last-Modified', None)
+                if last_modified:
+                    self._last_modified = last_modified
+
+                self._last_image = await res.read()
+                _LOG.debug("HTTP 200 - Last-Modified: %s", last_modified)
+
+                return True
+        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
+            _LOG.error("Failed to fetch image, %s", type(err))
+            return False
+
+    async def async_camera_image(self) -> Optional[bytes]:
+        """
+        Return a still image response from the camera.
+
+        Uses ayncio conditions to make sure only one task enters the critical
```

Comment:
I wonder if we need to make this a decorator that can be reused. We have similar logic inside Hue and probably will have it for other integrations that share a resource among requests.

Reply:
I considered this and will get back on this later (likely this weekend).

I see three scenario's here:
  * generator function used as coroutine, [seems](https://stackoverflow.com/questions/34469060/python-native-coroutines-and-send) to rule out async functions. I do not find this intuitive
  * a decorator. Advantage: does not depend on function name, does not need to be called explicitly. Tricky because it will take parameters (the time delta) from the instance of the class and this protocol is implicit.
  * an abstract base class with multiple inheritance. Advantage: Makes the lifecycle explicit. Disadvantage: more verbose

Reply:
I think I have found the generic/re-usable use case, will try to write down this design pattern later this week due to time constraints.

In the meantime I have made an assumption about lifecycle of integrations. The state returned is the state _at the time of the return_, instead of the time of calling? This would allow calls to be '"[folded](https://en.wikipedia.org/wiki/Fold_(higher-order_function)" to single calls in my design pattern.

This pattern can combine multiple "request update" calls into one, or combine multiple light switch actions into a single action performed on an integration (`turn on A`, `turn on B`, `turn on C` => `turn on (A, B, C,)`)

Reply:
State is at time of the return.

We don't do folded turn_on yet by platform. Right now we do allow platforms to specify max concurrency for calling services. States are only polled once all commands have been sent.